library/Freckle/App/Stats.hs

```diff
@@ -0,0 +1,280 @@
+        <*> Env.var Env.keyValues "DOGSTATSD_TAGS" (Env.def [])
+        )
+ where
+  buildSettings mHost mPort =
+    Datadog.defaultSettings
+      & maybe id (Datadog.host .~) mHost
+      . maybe id (Datadog.port .~) mPort
+
+  buildTags mEnv mService mVersion tags =
+    catMaybes
+        [ ("env", ) <$> mEnv
+        , ("environment", ) <$> mEnv -- Legacy
+        , ("service", ) <$> mService
+        , ("version", ) <$> mVersion
+        ]
+      <> tags
+
+data StatsClient = StatsClient
+  { scClient :: Datadog.StatsClient
+  , scTags :: [(Text, Text)]
+  }
+
+tagsL :: Lens' StatsClient [(Text, Text)]
+tagsL = lens scTags $ \x y -> x { scTags = y }
+
+class HasStatsClient env where
+  statsClientL :: Lens' env StatsClient
+
+instance HasStatsClient StatsClient where
+  statsClientL = id
+
+instance HasStatsClient site =>  HasStatsClient (HandlerData child site) where
+  statsClientL = envL . siteL . statsClientL
+
+withStatsClient
+  :: (MonadMask m, MonadUnliftIO m)
+  => StatsSettings
+  -> (StatsClient -> m a)
+  -> m a
+withStatsClient StatsSettings {..} f = do
+  if amsEnabled
+    then do
+      tags <- (amsTags <>) <$> getEcsMetadataTags
+
+      Datadog.withDogStatsD amsSettings $ \client ->
+        -- Add the tags to the thread context so they're present in all logs
+        withThreadContext (map toPair tags)
+          $ f StatsClient { scClient = client, scTags = tags }
+    else f $ StatsClient { scClient = Datadog.Dummy, scTags = [] }
+  where toPair = bimap (fromString . unpack) String
+
+-- | Include the given tags on all metrics emitted from a block
+tagged
+  :: (MonadReader env m, HasStatsClient env) => [(Text, Text)] -> m a -> m a
+tagged tags = local $ statsClientL . tagsL <>~ tags
+
+-- | Synonym for @'counter' 1@
+increment
+  :: (MonadUnliftIO m, MonadReader env m, HasStatsClient env) => Text -> m ()
+increment name = counter name 1
+
+counter
+  :: (MonadUnliftIO m, MonadReader env m, HasStatsClient env)
+  => Text
+  -> Int
+  -> m ()
+counter = sendMetric Datadog.Counter
+
+gauge
+  :: (MonadUnliftIO m, MonadReader env m, HasStatsClient env)
+  => Text
+  -> Double
+  -> m ()
+gauge = sendMetric Datadog.Gauge
+
+-- | Emit an elapsed duration (which Datadog calls a /histogram/)
+--
+-- The 'ToMetricValue' constraint can be satisfied by most numeric types and is
+-- assumed to be seconds.
+--
+histogram
+  :: ( MonadUnliftIO m
+     , MonadReader env m
+     , HasStatsClient env
+     , Datadog.ToMetricValue n
+     )
+  => Text
+  -> n
+  -> m ()
+histogram = sendMetric Datadog.Histogram
+
+histogramSince
+  :: (MonadUnliftIO m, MonadReader env m, HasStatsClient env)
+  => Text
+  -> UTCTime
+  -> m ()
+histogramSince = histogramSinceBy toSeconds where toSeconds = round @_ @Int
+
+histogramSinceMs
+  :: (MonadUnliftIO m, MonadReader env m, HasStatsClient env)
+  => Text
+  -> UTCTime
+  -> m ()
+histogramSinceMs = histogramSinceBy toMilliseconds
+  where toMilliseconds = (* 1000) . realToFrac @_ @Double
```

Comment:
Feeling a little [daft][daft]... Please walk me through why we multiply here instead of divide.

[daft]: https://www.youtube.com/watch?v=yydNF8tuVmU

Reply:
We already measure the duration in seconds. So if it took 5 seconds, we get `5` here and we want to report that as 5,000 ms, `5 * 1000`.

Reply:
This should probably be updated to use the monotonic stuff Evan used in Scientist, btw.

Reply:
Oh nice, gotcha!